### PR TITLE
Remove special processing of binary logical operators on CPU because bfloat16 type promotion is now available

### PR DESCRIPTION
**Pull request by xuhdev**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42485 Support BFloat16 for binary logical operators on CUDA
* **#42484 Remove special processing of binary logical operators on CPU because bfloat16 type promotion is now available**
* #42483 For logical tests, use the dtypes decorator

### CODE REVIEW

**Comment by peterbell10** [comment]
Is there any precedence for nested dispatching over two types at once? In fact I see the old version which special cased bool seems to be a pattern used throughout this file for comparison operations.

**Comment by xuhdev** [comment]
@peterbell10 I can find another nested instance at https://github.com/pytorch/pytorch/blob/7221a3d1aaa40eda0511f7f103bff67782313cf7/aten/src/ATen/native/cpu/CopyKernel.cpp#L43-L45

The important thing here is that these two (`if-else` vs nested dispatch) don't mean the same output when the return dtype is not bool and the two inputs have different dtypes. For example, the old code would produce:

```
In [1]: import torch

In [2]: a=torch.tensor([190, 220], dtype=torch.bfloat16)

In [3]: b=torch.tensor([190, 210], dtype=torch.float32)

In [4]: c=torch.empty(0, dtype=torch.float64)

In [5]: torch.logical_or(a, b, out=c)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-5-3ac99fe7cb2a> in <module>
----> 1 torch.logical_or(a, b, out=c)

RuntimeError: !needs_dynamic_casting<func_t>::check(iter) INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cpu/Loops.h":194, please report a bug to PyTorch. 
```

But the new code would go through without an error:

```
In [1]: import torch

In [2]: a=torch.tensor([190, 220], dtype=torch.bfloat16)

In [3]: b=torch.tensor([190, 210], dtype=torch.float32)

In [4]: c=torch.empty(0, dtype=torch.float64)

In [5]: torch.logical_or(a, b, out=c)
Out[5]: tensor([1., 1.], dtype=torch.float64)
```

**Comment by peterbell10** [comment]
Don't most operators cast to the common dtype so there is only O(n) kernels instead of O(n^2)? In which case this seems like a `TensorIterator` issue. The same pattern is used for all the comparison operators: `eq`, `ne`, `lt`, `le`, `gt`, `ge`. So, I presume these must all be broken in the same way.

**Comment by xuhdev** [comment]
@peterbell10 I don't think the nested dispatch would really affect runtime performance, because the number of dtypes available is bounded and really tiny compared with the operand. In theory, the change should drag compilation speed but in practice I observe no difference.

I think you are referring to the CUDA version of the implementation as the [next stacked PR](https://github.com/pytorch/pytorch/pull/42485/files) shows. The CPU version cannot duplicate that implementation because dynamic casting is not supported on CPU: https://github.com/pytorch/pytorch/blob/54ffb05eff27df4fdb3fc3f439851f55b86d57f1/aten/src/ATen/native/cpu/Loops.h#L193-L194
In other words, on CPU, you can't return a dtype that is different from the output tensor and expect TensorIterator to do the casting for you.

Yes, I did some experiments and the issue prevented in my previous comment is also applicable to comparison operators.

**Comment by peterbell10** [comment]
Oh, performance isn't my concern. In fact, this should be better for runtime performance than the normal way of casting tensors. It's just binary size that might be problematic. If you did this for every kernel, pytorch would balloon in size.

Other than that concern, this code looks good though.

**Comment by xuhdev** [comment]
So what is the current policy then? @ezyang Are you aware of any relevant policies on this?

**Comment by ezyang** [comment]
You better have a really good reason for nested dispatch, because, as @peterbell10 points out, you end up with quadratic code size when you do that. cc @malfet

**Comment by xuhdev** [comment]
Sounds good. I'm dropping this PR from the stack and will try to come up with a solution for all comparison operators.

**Comment by xuhdev** [comment]
I noticed `logical_not` also nests dispatches and has explained it with comments:

https://github.com/pytorch/pytorch/blob/d5bc2a8058c79d3a2d5fc9c9a186a78de3342d70/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp#L233-L243

So... does this really fit in our policy?

**Comment by peterbell10** [comment]
Two interesting points:
- The double dispatch was only added to `logical_not` fairly recently by yourself in #23916
- #38565 removed the double dispatch from CUDA, citing compilation speed.
